### PR TITLE
centroid_triplet_loss implementation

### PR DESCRIPTION
**Pull request by codeandproduce**
Testing script specifically tests additionally for:

- Imbalanced mini-batches 
- Returning same results regardless of the ordering of embeddings/labels in the mini-batches

### CODE REVIEW

**Comment by KevinMusgrave** [comment]
Thanks for working on this. I'm busy preparing for a paper deadline, so once that's over I'll take a look and merge.

**Comment by KevinMusgrave** [comment]
When I run ```test_centroid_triplet_loss``` I get the following embeddings:
```python
tensor([[1.0000, 0.0000],
        [0.9848, 0.1736],
        [0.9397, 0.3420],
        [0.8660, 0.5000],
        [0.7660, 0.6428],
        [0.6428, 0.7660]])
```

labels: ```torch.LongTensor([0, 0, 0, 1, 1, 1])```

centroids:
```python
tensor([[[0.9623, 0.2578],
         [0.7044, 0.7044]],

        [[0.9698, 0.1710],
         [0.7544, 0.6330]],

        [[0.9924, 0.0868],
         [0.8160, 0.5714]]])
```

The first triplet indices are ```0 (0, 0) (0, 1)```, which correspond with:

anchor: ```tensor([1., 0.]```
positive centroid: ```tensor([0.9623, 0.2578]```
negative centroid: ```tensor([0.7044, 0.7044]```

The positive centroid is the average of embeddings 1 and 2, and the anchor is embedding 0, so that makes sense. But the negative centroid seems to be the average of embeddings 5 and 6, rather than 4, 5, and 6. Is there a reason why embedding #4 is excluded?

**Comment by codeandproduce** [comment]
I made it this way to make it fit nicely with `lmu.get_all_triplets_indices`:

Consider the average of embeddings 1 and 2, which is the **positive centroid** for the anchor 0 as you mentioned.
but with the current implementation it can be the **negative centroid** for the embedding 3.

In this case:
anchor: `0`
positive: `avg(1, 2)`
negative: `avg(4, 5)`

anchor: `3`
positive: `avg(4, 5)`
negative: `avg(1, 2)`

which fits nicely with the use of `lmu.get_all_triplets_indices`:
```python
embeddings = [[1.0000, 0.0000], [0.8660, 0.5000]]
labels = [0, 1]

centroids = [
   [0.9623, 0.2578] # avg(1, 2)
   [0.7044, 0.7044] # avg(4, 5)
]

embeddings = torch.cat((embeddings, centroids))
labels = torch.cat((labels, labels))
triples = lmu.get_all_triplets_indices(labels)

# then remove triples where the centroid vector ends up being the anchor vector
```

A more principled approach would be:
```python
embeddings = [
    [1.0000, 0.0000],  # embedding 0
    [0.8660, 0.5000],  # embedding 3
    [0.9623, 0.2578],   # positive for embedding 0
    [0.7044, 0.7044],   # positive for embdding 3
    [0.7583, 0.6363],   # negative for embedding 0 == avg(3,4,5)
    [0.9748, 0.1719]   # negative for embedding 3 == avg(0,1,2)
]
triples = [
[0, 2, 4],
[1, 3, 5]
]
```

Thinking back, considering that there isn't that much more computational overhead, the latter seems like the more correct approach. I guess I was thinking of how much neater the former is and how losing one embedding in a centroid calculation makes minimal difference if not a regularization effect preventing it from memorizing the same centroids that would be used across the full batch pass.

Let me know what you think! If you obviously think the latter is the better approach, I could modify the implementation and submit a new pull request.

**Comment by KevinMusgrave** [comment]
Yes I think it's better to use the true centroid since that is what the user would probably expect.

**Comment by codeandproduce** [comment]
I've finished the above "true centroids" implementation. 

The tests uses the following method: 
```python
embedding_angles = [0, 10, 20, 30, 40, 50]
centroid_makers = [
    [[10, 20], [30, 40, 50]],
    [[0, 20], [30, 40, 50]],
    [[0, 10], [30, 40, 50]],
    [[40, 50], [0, 10, 20]],
    [[30, 50], [0, 10, 20]],
    [[30, 40], [0, 10, 20]]
]
triplets = [
    (0, (0, 0), (0, 1)),
    (1, (1, 0), (1, 1)),
    (2, (2, 0), (2, 1)),
    (3, (3, 0), (3, 1)),
    (4, (4, 0), (4, 1)),
    (5, (5, 0), (5, 1))
]
```

`centroid_makers` is made so that each item represents the list of angles whose mean embeddings will form that centroid.
```
centroid_makers[0][0] = positive centroid for embedding 0
centroid_makers[0][1] = negative centroid for embedding 0
centroid_makers[1][0] = positive centroid for embedding 1

```

Let me know what you think!

**Comment by KevinMusgrave** [comment]
The only change I made (other than formatting) was getting rid of the triplet loss attributes (self.margin, self.smooth_loss etc.) from CentroidTripletLoss, because they don't reflect the values inside self.triplet_loss. Someone might change them from the outside and expect them to also change in self.triplet_loss.

Other than that, it looks good to me! Thanks for working on this, I appreciate it! The loss function and tests look like they must have been very challenging to figure out.
